File: Task.java

import java.time.*;

/**Task class which can have child tasks and data, such as
* the task's name, the project, the task ID, and names and
* IDs of the employees who work on the task associated with it
*/
public class Task extends Activity {
    private int baseProjectID;

    public Task() {
        this.setName("Unnamed Task");
    }

    public Task(String newTaskName) {
        super(newTaskName);
    }

    public Task(String newProjectName, LocalDate newDeadline) {
        super(newProjectName, newDeadline);
    }

    

    //Getters and setters

    /**
     * @param newBaseProjectID The ID of the project the task belongs to
     */
    public void setactivityProject(int newBaseProjectID) {
        this.baseProjectID = newBaseProjectID;
    }

    /**
     * @return baseProjectID The ID of the project the task belongs to
     */
    public int getBaseProject() {
        return this.baseProjectID;
    }
}

File: models/Task.js
const Activity = require('./Activity'); // Import Activity base class
const pool = require('../config/db'); // Import database connection

class Task extends Activity {
    constructor(newTaskName = "Unnamed Task", newProjectName = null, newDeadline = null) {
        super(newTaskName, null, newDeadline); // Call Activity constructor
        this.baseProjectID = null; // Link to a parent project
    }

    // Getter and setter for the base project ID
    setActivityProject(newBaseProjectID) {
        this.baseProjectID = newBaseProjectID;
    }

    getBaseProject() {
        return this.baseProjectID;
    }

    // Save the task to the database
    async save() {
        const sql = 'INSERT INTO Tasks (name, startDate, deadline, baseProjectID) VALUES (?, ?, ?, ?)';
        const [result] = await pool.query(sql, [this.name, this.startDate, this.deadline, this.baseProjectID]);
        this.id = result.insertId; // Assign the generated database ID
    }
}

module.exports = Task; // Export the class
